url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/SessionImpl.java

```diff
@@ -0,0 +1,1130 @@
+                Listener listener )
+    {
+        if (listener == null)
+            closeConnection();
+        else
+        {
+            /* m_socketChannelReader should be initialized before state change
+             * to the ST_RUNNING, closeConnection() method relies on it.
+             */
+            m_socketChannelReader = new SocketChannelReader(
+                    m_collider,
+                    this,
+                    inputQueueMaxSize,
+                    inputQueueDataBlockCache,
+                    m_socketChannel,
+                    m_selectionKey,
+                    listener );
+        }
+
+        for (;;)
+        {
+            final int state = m_state.get();
+            assert( (state & STATE_MASK) == ST_STARTING );
+            if ((state & CLOSE) == 0)
+            {
+                assert( (state & SOCK_RC_MASK) == SOCK_RC );
+                int newState = state;
+                newState &= ~STATE_MASK;
+                newState |= ST_RUNNING;
+                newState += SOCK_RC;
+                if (m_state.compareAndSet(state, newState))
+                {
+                    if (s_logger.isLoggable(Level.FINE))
+                    {
+                        s_logger.fine(
+                                m_localSocketAddress + " -> " + m_remoteSocketAddress +
+                                ": " + stateToString(state) + " -> " + stateToString(newState) + "." );
+                    }
+                    m_socketChannelReader.start();
+                    break;
+                }
+            }
+            else
+            {
+                if (s_logger.isLoggable(Level.FINE))
+                {
+                    s_logger.fine(
+                            m_localSocketAddress + " -> " + m_remoteSocketAddress +
+                            ": " + stateToString(state) + "." );
+                }
+
+                if (m_socketChannelReader != null)
+                {
+                    /* listener != null */
+                    m_socketChannelReader.reset();
+                }
+
+                break;
+            }
+        }
+    }
+
+    public Collider getCollider() { return m_collider; }
+    public SocketAddress getLocalAddress() { return m_localSocketAddress; }
+    public SocketAddress getRemoteAddress() { return m_remoteSocketAddress; }
+
+    public int sendData( ByteBuffer data )
+    {
+        assert( data.remaining() > 0 );
+        final Node node = new Node( data );
+        for (;;)
+        {
+            final Node tail = m_tail.get();
+            if (tail == CLOSE_MARKER)
+                return -1;
+
+            if (m_tail.compareAndSet(tail, node))
+            {
+                if (tail == null)
+                {
+                    m_head = node;
+                    m_collider.executeInThreadPool( m_writer );
+                }
+                else
+                    tail.next = node;
+                return 1;
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 746:763, 770:787.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.